### PR TITLE
Fix mapping for `tpp.identity`

### PR DESCRIPTION
**Pull request by chelini**
The mapping logic for `tpp.identity` was too naive. So now we check the affine maps and require them to be projected permutation. Additionally, we ensure before mapping that the linalg op verifies "our" broadcasting rules and rejects candidate ops that do not satisfy these requirements. Current mapping is still not optimal, and more effort needs to be spent on mapping and broadcast support.

### CODE REVIEW

File: lib/TPP/Dialect/Tpp/TppUtils.cpp [hunk 1]
```diff
@@ -355,6 +362,39 @@ static bool hasMaxfZeroOp(linalg::LinalgOp linalgOp) {
   return false;
 }
 
+MatchBroadcastRuleResult verifyTppIdentityBroadcastingRules(Type inputType,
+                                                            Type outputType) {
+  // input scalar, just return.
+  if (!inputType.isa<ShapedType>())
```
Comment (rengolin) [inline lib/TPP/Dialect/Tpp/TppUtils.cpp:368]: common up `inputType.cast<ShapedType>()`

File: lib/TPP/Dialect/Tpp/TppUtils.cpp [hunk 2]
```diff
@@ -363,8 +403,7 @@ bool isTppAdd(linalg::GenericOp linalgOp) {
     return false;
   if (!allOperandsHaveSameShapeAndStrides(linalgOp->getOperands().getTypes()))
     return false;
-  return allIndexingsAreProjectedPermutation(linalgOp) &&
-         hasOnlyOp<arith::AddFOp>(linalgOp.getRegion());
+  return hasOnlyOp<arith::AddFOp>(linalgOp.getRegion());
```
Comment (rengolin) [inline lib/TPP/Dialect/Tpp/TppUtils.cpp:406]: Why this change?

File: lib/TPP/Dialect/Tpp/TppUtils.cpp [hunk 3]
```diff
@@ -385,7 +423,12 @@ bool isTppIdentity(linalg::GenericOp linalgOp) {
     return false;
   if (!isUnaryOp(linalgOp))
     return false;
-  return hasCopySemantics(linalgOp);
+  if (!hasCopySemantics(linalgOp))
+    return false;
+  auto res = verifyTppIdentityBroadcastingRules(
+      linalgOp.getDpsInputOperand(0)->get().getType(),
```
Comment (rengolin) [inline lib/TPP/Dialect/Tpp/TppUtils.cpp:429]: `linalgOp.getOperand(0).getType()`?

File: test/Conversion/LinalgToTpp/linalg-to-tpp.mlir [hunk 4]
```diff
@@ -556,3 +556,171 @@ func.func @relu_mapping_strided(%i: memref<4x4xf32>, %o: memref<4x4xf32, strided
...
+
+// -----
+
+#map = affine_map<(d0, d1) -> ()>
+#map1 = affine_map<(d0, d1) -> (d0, d1)>
+
+// CHECK-LABEL: func.func @empty_map_identity
+// CHECK-SAME: %[[ARG0:.+]]: f32, %[[ARG1:.+]]: memref<8x32xf32>
+func.func @empty_map_identity(%arg0 : f32, %arg1: memref<8x32xf32>) {
+  // CHECK: ins(%[[ARG0]] : f32) out(%[[ARG1]] : memref<8x32xf32>)
```
Comment (rengolin) [inline test/Conversion/LinalgToTpp/linalg-to-tpp.mlir:642]: What about tpp.identity? No checks?

**Comment by rengolin** [comment]
This PR needs to also fix the MHA code in #342

File: lib/TPP/Dialect/Tpp/TppUtils.cpp (hunk 2 above)
Comment (chelini) [reply lib/TPP/Dialect/Tpp/TppUtils.cpp:406]: We use projectedPermutation for each tpp. So now we check directly in `hasMappingToTppConditions`

File: test/Conversion/LinalgToTpp/linalg-to-tpp.mlir (hunk 4 above)
Comment (chelini) [reply test/Conversion/LinalgToTpp/linalg-to-tpp.mlir]: forgot while typing.

**Review by rengolin — APPROVED** [review]
Just tested on my MLP branch and it works (ie. doesn't match the wrong generic as a tpp.identity), so I'm happy.

The MHA also seems to work, but the checks are now different and the PR may need updating. @nhasabni make sure that's correct.
